subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/result/DefaultComponentSelectionDescriptor.java

```diff
@@ -25,19 +25,22 @@
     private final Describable description;
     private final boolean hasCustomDescription;
     private final int hashCode;
+    private final boolean isEquivalentToForce;
 
     public DefaultComponentSelectionDescriptor(ComponentSelectionCause cause) {
-        this.cause = cause;
-        this.description = Describables.of(cause.getDefaultReason());
-        this.hasCustomDescription = false;
-        this.hashCode = cause.hashCode();
+        this(cause, Describables.of(cause.getDefaultReason()), false, cause.hashCode(), cause == ComponentSelectionCause.FORCED);
     }
 
     public DefaultComponentSelectionDescriptor(ComponentSelectionCause cause, Describable description) {
+        this(cause, description, true, Objects.hashCode(cause, description, cause == ComponentSelectionCause.FORCED), cause == ComponentSelectionCause.FORCED);
+    }
+
+    private DefaultComponentSelectionDescriptor(ComponentSelectionCause cause, Describable description, boolean hasCustomDescription, int hashCode, boolean isEquivalentToForce) {
         this.cause = cause;
         this.description = description;
-        this.hasCustomDescription = true;
-        this.hashCode = Objects.hashCode(cause, description);
+        this.hasCustomDescription = hasCustomDescription;
+        this.hashCode = hashCode;
```

Comment:
:o: It feels a bit strange to me to provide the hash code _externally_ to the constructor.

Reply:
Found a better way of doing this, that no longer requires it to be a constructor parameter.